apis/v1alpha2/httproute_types.go

```diff
@@ -408,6 +408,25 @@ type HTTPQueryParamMatch struct {
 	Value string `json:"value"`
 }
 
+// HTTPMethod describes how to select a HTTP route by matching the HTTP
+// method as defined by [RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-4) and [RFC 5789](https://datatracker.ietf.org/doc/html/rfc5789#section-2).
+// The value is expected in upper case
+// +kubebuilder:validation:Enum=GET;HEADER;POST;PUT;DELETE;CONNECT;OPTIONS;TRACE;PATCH
+type HTTPMethod string
+
+const (
+	HTTPMethodUndefined HTTPMethod = "NONE"
+	HTTPMethodGet       HTTPMethod = "GET"
+	HTTPMethodHeader    HTTPMethod = "HEADER"
+	HTTPMethodPost      HTTPMethod = "POST"
+	HTTPMethodPut       HTTPMethod = "PUT"
+	HTTPMethodDelete    HTTPMethod = "DELETE"
+	HTTPMethodConnect   HTTPMethod = "CONNECT"
+	HTTPMethodOptions   HTTPMethod = "OPTIONS"
+	HTTPMethodTrace     HTTPMethod = "TRACE"
+	HTTPMethodPatch     HTTPMethod = "PATCH"
```

Comment:
I'm not sure if this is a good idea or not.
Custom HTTP methods are a thing, although not common in practice.
For the 99% of the cases, these defined methods will suffice.
If there comes a time when we want to support custom HTTP methods, how can we do so without a breaking change?

Reply:
> If I'm not mistaken it is already possible to match the HTTP method via header value match.

How so?


cc @robscott @jpeach

Reply:
With `:method` pseudo-header (https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.1)

If we go down this route we need to make it explicit in the spec that implementation must accept pseudo-headers though.

Reply:
> If we go down this route we need to make it explicit in the spec that implementation must accept pseudo-headers though.

We should clarify this regardless. Else implementations will drift in this aspect.

Reply:
Yup, +1 on defining pseudo-header matching. Controllers can map the pseudo header name to the underlying proxy control.

Reply:
I opened up https://github.com/kubernetes-sigs/gateway-api/issues/740 to track this conversation.

My only slight hesitation is that then there are two ways to do the same thing. In this case, that might be okay because custom HTTP methods are not common. Not really sure.